src/coprocessor/readpool_impl.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cell::RefCell;
+
+use crate::pd::PdTask;
+use crate::server::readpool::{self, Builder, ReadPool};
+use crate::util::collections::HashMap;
+use crate::util::worker::FutureScheduler;
+
+use super::metrics::*;
+use prometheus::local::*;
+
+use crate::coprocessor::dag::executor::ExecutorMetrics;
+
+pub struct TlsCop {
+    pub local_copr_req_histogram_vec: RefCell<LocalHistogramVec>,
+    pub local_outdated_req_wait_time: RefCell<LocalHistogramVec>,
+    pub local_copr_req_handle_time: RefCell<LocalHistogramVec>,
+    pub local_copr_req_wait_time: RefCell<LocalHistogramVec>,
+    pub local_copr_req_error: RefCell<LocalIntCounterVec>,
+    pub local_copr_scan_keys: RefCell<LocalHistogramVec>,
+    pub local_copr_scan_details: RefCell<LocalIntCounterVec>,
+    pub local_copr_rocksdb_perf_counter: RefCell<LocalIntCounterVec>,
+    local_copr_executor_count: RefCell<LocalIntCounterVec>,
+    local_copr_get_or_scan_count: RefCell<LocalIntCounterVec>,
+    local_cop_flow_stats: RefCell<HashMap<u64, crate::storage::FlowStatistics>>,
```

Comment:
There are too many `RefCells`. How about put the struct in a `RefCell` or `Mutex`? I think it's more clear.

Reply:
Nice catch. You should arrange them as..

```rust
pub struct Xxx {
    field: LocalIntCounter,
    field_2: LocalIntCounter,
    ...
}

thread_local! {
    pub static TLS_COP_METRICS: RefCell<TlsCop> = ...;
}
```

In this way, we only need to check borrow once when updating multiple fields.